File: src/main/java/ltdjms/discord/gametoken/services/GameTokenService.java
package ltdjms.discord.gametoken.services;

import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.gametoken.domain.GameTokenAccount;
import ltdjms.discord.gametoken.persistence.GameTokenAccountRepository;
import ltdjms.discord.gametoken.persistence.InsufficientTokensException;
import ltdjms.discord.shared.DomainError;
import ltdjms.discord.shared.Result;
import ltdjms.discord.shared.cache.CacheKeyGenerator;
import ltdjms.discord.shared.cache.CacheService;
import ltdjms.discord.shared.events.DomainEventPublisher;
import ltdjms.discord.shared.events.GameTokenChangedEvent;

/**
 * Service for managing game token accounts. Handles token adjustments with validation for
 * non-negative balances.
 */
public class GameTokenService {

  private static final Logger LOG = LoggerFactory.getLogger(GameTokenService.class);
  private static final int TOKEN_TTL_SECONDS = 300;

  private final GameTokenAccountRepository accountRepository;
  private final DomainEventPublisher eventPublisher;
  private final CacheService cacheService;
  private final CacheKeyGenerator cacheKeyGenerator;

  public GameTokenService(
      GameTokenAccountRepository accountRepository,
      DomainEventPublisher eventPublisher,
      CacheService cacheService,
      CacheKeyGenerator cacheKeyGenerator) {
    this.accountRepository = accountRepository;
    this.eventPublisher = eventPublisher;
    this.cacheService = cacheService;
    this.cacheKeyGenerator = cacheKeyGenerator;
  }

  /**
   * Gets the current token balance for a member.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @return the token balance (0 if account doesn't exist)
   */
  public long getBalance(long guildId, long userId) {
    Long cachedBalance = getCachedBalance(guildId, userId);
    if (cachedBalance != null) {
      return cachedBalance;
    }

    // Cache miss - get from database
    long balance =
        accountRepository
            .findByGuildIdAndUserId(guildId, userId)
            .map(GameTokenAccount::tokens)
            .orElse(0L);
    putCachedBalance(guildId, userId, balance);
    return balance;
  }

  /**
   * Adjusts a member's token balance by the specified amount.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param amount the amount to adjust (positive for credit, negative for debit)
   * @return the adjustment result
   * @throws InsufficientTokensException if the adjustment would result in a negative balance
   */
  public TokenAdjustmentResult adjustTokens(long guildId, long userId, long amount) {
    LOG.debug("Adjusting tokens for guildId={}, userId={}, amount={}", guildId, userId, amount);

    // Get current balance
    GameTokenAccount current = accountRepository.findOrCreate(guildId, userId);
    long previousTokens = current.tokens();

    // Apply adjustment
    GameTokenAccount updated = accountRepository.adjustTokens(guildId, userId, amount);

    // Update cache
    updateCachedBalance(guildId, userId, updated.tokens());

    // Publish event
    eventPublisher.publish(new GameTokenChangedEvent(guildId, userId, updated.tokens()));

    TokenAdjustmentResult result =
        new TokenAdjustmentResult(guildId, userId, previousTokens, updated.tokens(), amount);

    LOG.info(
        "Tokens adjusted: guildId={}, userId={}, previous={}, new={}, adjustment={}",
        guildId,
        userId,
        previousTokens,
        updated.tokens(),
        amount);

    return result;
  }

  /**
   * Checks if a member has enough tokens.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param requiredTokens the number of tokens required
   * @return true if the member has enough tokens
   */
  public boolean hasEnoughTokens(long guildId, long userId, long requiredTokens) {
    return getBalance(guildId, userId) >= requiredTokens;
  }

  /**
   * Deducts tokens from a member's account.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param tokens the number of tokens to deduct (must be positive)
   * @return the updated account
   * @throws IllegalArgumentException if tokens is not positive
   * @throws InsufficientTokensException if the member doesn't have enough tokens
   */
  public GameTokenAccount deductTokens(long guildId, long userId, long tokens) {
    if (tokens <= 0) {
      throw new IllegalArgumentException("Tokens to deduct must be positive: " + tokens);
    }
    GameTokenAccount updated = accountRepository.adjustTokens(guildId, userId, -tokens);

    // Update cache
    updateCachedBalance(guildId, userId, updated.tokens());

    // Publish event
    eventPublisher.publish(new GameTokenChangedEvent(guildId, userId, updated.tokens()));

    return updated;
  }

  /**
   * Adjusts a member's token balance using Result-based error handling.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param amount the amount to adjust (positive for credit, negative for debit)
   * @return Result containing TokenAdjustmentResult on success, or DomainError on failure
   */
  public Result<TokenAdjustmentResult, DomainError> tryAdjustTokens(
      long guildId, long userId, long amount) {
    LOG.debug("Adjusting tokens for guildId={}, userId={}, amount={}", guildId, userId, amount);

    // Get current balance
    GameTokenAccount current = accountRepository.findOrCreate(guildId, userId);
    long previousTokens = current.tokens();

    // Apply adjustment using Result-based API
    Result<GameTokenAccount, DomainError> adjustResult =
        accountRepository.tryAdjustTokens(guildId, userId, amount);

    if (adjustResult.isErr()) {
      return Result.err(adjustResult.getError());
    }

    GameTokenAccount updated = adjustResult.getValue();

    // Update cache
    updateCachedBalance(guildId, userId, updated.tokens());

    // Publish event
    eventPublisher.publish(new GameTokenChangedEvent(guildId, userId, updated.tokens()));

    TokenAdjustmentResult result =
        new TokenAdjustmentResult(guildId, userId, previousTokens, updated.tokens(), amount);

    LOG.info(
        "Tokens adjusted: guildId={}, userId={}, previous={}, new={}, adjustment={}",
        guildId,
        userId,
        previousTokens,
        updated.tokens(),
        amount);

    return Result.ok(result);
  }

  /**
   * Deducts tokens using Result-based error handling.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param tokens the number of tokens to deduct (must be positive)
   * @return Result containing the updated account, or DomainError on failure
   */
  public Result<GameTokenAccount, DomainError> tryDeductTokens(
      long guildId, long userId, long tokens) {
    if (tokens <= 0) {
      return Result.err(DomainError.invalidInput("Tokens to deduct must be positive: " + tokens));
    }
    Result<GameTokenAccount, DomainError> result =
        accountRepository.tryAdjustTokens(guildId, userId, -tokens);

    if (result.isOk()) {
      // Update cache
      updateCachedBalance(guildId, userId, result.getValue().tokens());

      // Publish event
      eventPublisher.publish(
          new GameTokenChangedEvent(guildId, userId, result.getValue().tokens()));
    }

    return result;
  }

  /** Result of a token adjustment operation. */
  public record TokenAdjustmentResult(
      long guildId, long userId, long previousTokens, long newTokens, long adjustment) {
    /** Formats the result as a Discord message. */
    public String formatMessage(String targetUserMention) {
      String action = adjustment >= 0 ? "Added" : "Removed";
      long displayAmount = Math.abs(adjustment);
      return String.format(
          "%s %,d game tokens to/from %s\nNew balance: %,d game tokens",
          action, displayAmount, targetUserMention, newTokens);
    }
  }

  /**
   * Gets the token balance from cache.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @return the cached balance, or null if not in cache
   */
  private Long getCachedBalance(long guildId, long userId) {
    String cacheKey = cacheKeyGenerator.gameTokenKey(guildId, userId);
    Optional<Long> cached = cacheService.get(cacheKey, Long.class);
    if (cached.isPresent()) {
      LOG.debug("Cache hit for game token: guildId={}, userId={}", guildId, userId);
      return cached.get();
    }
    LOG.debug("Cache miss for game token: guildId={}, userId={}", guildId, userId);
    return null;
  }

  /**
   * Puts the token balance into cache.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param balance the balance to cache
   */
  private void putCachedBalance(long guildId, long userId, long balance) {
    String cacheKey = cacheKeyGenerator.gameTokenKey(guildId, userId);
    cacheService.put(cacheKey, balance, TOKEN_TTL_SECONDS);
    LOG.debug("Cached game token: guildId={}, userId={}, balance={}", guildId, userId, balance);
  }

  /**
   * Updates the cached token balance.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param balance the new balance to cache
   */
  private void updateCachedBalance(long guildId, long userId, long balance) {
    String cacheKey = cacheKeyGenerator.gameTokenKey(guildId, userId);
    cacheService.put(cacheKey, balance, TOKEN_TTL_SECONDS);
    LOG.debug(
        "Updated cached game token: guildId={}, userId={}, balance={}", guildId, userId, balance);
  }
}

File: packages/economy/src/token/services/game-token-service.ts
import {
  type Result,
  Ok,
  Err,
  DomainError,
  type CacheService,
  type CacheKeyGenerator,
  type DomainEventPublisher,
} from '@ltdjms/shared';
import type { GameTokenChangedEvent } from '@ltdjms/economy';
import { TokenAccountRepository } from '../repositories/token-account-repo.js';
import { GameTokenTransactionService } from './game-token-tx-service.js';
import type { GameTokenAccount, TokenAdjustmentResult } from '../../domain/types.js';
import { TOKEN_CACHE_TTL, GameTokenTransactionSource } from '../../domain/types.js';

/**
 * Service for managing game token accounts with caching.
 * Matches Java GameTokenService behavior.
 */
export class GameTokenService {
  private static readonly TOKEN_TTL_SECONDS = TOKEN_CACHE_TTL;

  /** Per-key in-flight promises to prevent cache stampede on token reads. */
  private readonly pendingFetches = new Map<string, Promise<number>>();

  /**
   * Updates the cache and publishes a GameTokenChangedEvent after a token adjustment.
   * Extracted to eliminate duplicate cache/event logic across four methods (P1-12).
   */
  private async updateCacheAndPublishEvent(
    guildId: number,
    userId: number,
    newTokens: number,
  ): Promise<void> {
    const cacheKey = this.cacheKeyGenerator.gameTokenKey(String(guildId), String(userId));
    await this.cacheService.put(
      cacheKey,
      newTokens,
      GameTokenService.TOKEN_TTL_SECONDS,
    );

    const event: GameTokenChangedEvent = {
      guildId: String(guildId),
      userId,
      eventType: 'game_token_changed',
      newTokens,
    };
    this.eventPublisher.publish(event);
  }

  constructor(
    private readonly accountRepository: TokenAccountRepository,
    private readonly eventPublisher: DomainEventPublisher,
    private readonly cacheService: CacheService,
    private readonly cacheKeyGenerator: CacheKeyGenerator,
    private readonly transactionService: GameTokenTransactionService,
  ) {}

  /**
   * Gets the current token balance for a member.
   * Uses cache (TTL 300s) - cache miss falls through to DB query.
   * Auto-creates the token account if it does not exist (P3-16).
   */
  async getBalance(guildId: number, userId: number): Promise<number> {
    const cacheKey = this.cacheKeyGenerator.gameTokenKey(String(guildId), String(userId));
    const cachedBalance = await this.cacheService.get<number>(cacheKey);

    if (cachedBalance !== null) {
      return cachedBalance;
    }

    // Prevent cache stampede: coalesce concurrent requests for the same key
    const pending = this.pendingFetches.get(cacheKey);
    if (pending) {
      return await pending;
    }

    const fetchPromise = this.accountRepository.findOrCreate(guildId, userId)
      .then(account => account.tokens);
    this.pendingFetches.set(cacheKey, fetchPromise);
    try {
      const balance = await fetchPromise;
      await this.cacheService.put(cacheKey, balance, GameTokenService.TOKEN_TTL_SECONDS);
      return balance;
    } finally {
      this.pendingFetches.delete(cacheKey);
    }
  }

  /**
   * Adjusts tokens with Result-based error handling.
   * Records a transaction after the adjustment, cache update, and event publishing (P1-5).
   *
   * @param source - transaction source for recording (default ADMIN_ADJUSTMENT)
   * @param description - optional description for the transaction record
   */
  async tryAdjustTokens(
    guildId: number,
    userId: number,
    amount: number,
    source: GameTokenTransactionSource = GameTokenTransactionSource.ADMIN_ADJUSTMENT,
    description: string | null = null,
  ): Promise<Result<TokenAdjustmentResult, DomainError>> {
    if (!Number.isFinite(amount)) {
      return new Err(
        DomainError.invalidInput(`Invalid adjustment amount: ${amount}`),
      );
    }

    if (amount === 0) {
      return new Err(
        DomainError.invalidInput('調整金額不可為零'),
      );
    }

    try {
      const current = await this.accountRepository.findOrCreate(guildId, userId);
      const previousTokens = current.tokens;

      const adjustResult = await this.accountRepository.tryAdjustTokens(
        guildId,
        userId,
        amount,
      );

      if (adjustResult.isErr()) {
        return new Err(adjustResult.getError());
      }

      const updated = adjustResult.getValue();

      // Update cache and publish event
      await this.updateCacheAndPublishEvent(guildId, userId, updated.tokens);

      // Record transaction atomically with the token adjustment (P1-5)
      await this.transactionService.recordTransaction(
        guildId,
        userId,
        amount,
        updated.tokens,
        source,
        description,
      );

      return new Ok({
        guildId,
        userId,
        previousTokens,
        newTokens: updated.tokens,
        adjustment: amount,
      });
    } catch (err) {
      return new Err(
        DomainError.persistenceFailure(
          `Failed to adjust tokens for guildId=${guildId}, userId=${userId}`,
          err instanceof Error ? err : undefined,
        ),
      );
    }
  }

  /**
   * Checks if a member has enough tokens.
   */
  async hasEnoughTokens(
    guildId: number,
    userId: number,
    requiredTokens: number,
  ): Promise<boolean> {
    const balance = await this.getBalance(guildId, userId);
    return balance >= requiredTokens;
  }

  /**
   * Deducts tokens using Result-based error handling, then immediately records
   * the transaction so that deduction and recording are within the same logical
   * scope (P1-10). This prevents the process from crashing between the two
   * operations.
   *
   * Delegates to {@link tryAdjustTokens} for the actual adjustment, cache
   * update, and event publishing to avoid duplicating that logic (P2-16).
   *
   * @param tokens - positive number of tokens to deduct
   * @param source - transaction source for recording
   */
  async tryDeductTokens(
    guildId: number,
    userId: number,
    tokens: number,
    source: GameTokenTransactionSource = GameTokenTransactionSource.GAME_PLAY,
  ): Promise<Result<TokenAdjustmentResult, DomainError>> {
    if (tokens <= 0) {
      return new Err(
        DomainError.invalidInput(`Tokens to deduct must be positive: ${tokens}`),
      );
    }

    // Delegate to tryAdjustTokens which handles findOrCreate, adjustment,
    // cache update, event publishing, and transaction recording (P1-5, P2-16)
    const result = await this.tryAdjustTokens(guildId, userId, -tokens, source, null);

    return result;
  }

  /**
   * Adjusts tokens (deducts if negative, adds if positive).
   * Deduct path that throws on insufficient - use tryAdjustTokens for Result-based.
   * Records a transaction after successful deduction (P1-9).
   */
  async deductTokens(
    guildId: number,
    userId: number,
    tokens: number,
  ): Promise<GameTokenAccount> {
    if (tokens <= 0) {
      throw new Error(`Tokens to deduct must be positive: ${tokens}`);
    }

    const updated = await this.accountRepository.adjustTokens(
      guildId,
      userId,
      -tokens,
    );

    // Update cache and publish event
    await this.updateCacheAndPublishEvent(guildId, userId, updated.tokens);

    // Record transaction after successful deduction (P1-9), matching
    // the pattern used by tryDeductTokens (P1-10).
    await this.transactionService.recordTransaction(
      guildId,
      userId,
      -tokens,
      updated.tokens,
      GameTokenTransactionSource.GAME_PLAY,
      null,
    );

    return updated;
  }
}
